protocol/testing/containertest/testnet_test.go

```diff
@@ -262,3 +275,60 @@ func TestMarketPrices(t *testing.T) {
 	}
 	assertPricesWithTimeout(t, node, expectedPrices, 30*time.Second)
 }
+
+func TestUpgrade(t *testing.T) {
+	testnet, err := NewTestnetWithPreupgradeGenesis()
+	require.NoError(t, err, "failed to create testnet - is docker daemon running?")
+	err = testnet.Start()
+	require.NoError(t, err)
+	defer testnet.MustCleanUp()
+	node := testnet.Nodes["alice"]
+
+	proposal, err := gov.NewMsgSubmitProposal(
+		[]sdk.Msg{
+			&upgrade.MsgSoftwareUpgrade{
+				Authority: govModuleAddress,
+				Plan: upgrade.Plan{
+					Name:   UpgradeToVersion,
+					Height: 10,
+				},
+			},
+		},
+		testapp.TestDeposit,
+		constants.AliceAccAddress.String(),
+		testapp.TestMetadata,
+		testapp.TestTitle,
+		testapp.TestSummary,
+	)
+	require.NoError(t, err)
+
+	require.NoError(t, BroadcastTx(
+		node,
+		proposal,
+		constants.AliceAccAddress.String(),
+	))
+	err = node.Wait(2)
+	require.NoError(t, err)
+
+	addresses := []string{
+		constants.AliceAccAddress.String(),
+		constants.BobAccAddress.String(),
+		constants.CarlAccAddress.String(),
+		constants.DaveAccAddress.String(),
+	}
+
+	for _, address := range addresses {
+		require.NoError(t, BroadcastTx(
+			node,
+			&gov.MsgVote{
+				ProposalId: 1,
+				Voter:      address,
+				Option:     gov.VoteOption_VOTE_OPTION_YES,
+			},
+			address,
+		))
+	}
+
+	err = node.WaitUntilBlockHeight(12)
+	require.NoError(t, err)
```

Comment:
Will we run other "baseline" tests after upgrade, like placing order, etc?

Reply:
Not for now. I don't think it is that helpful for us because we'd need 100% coverage to catch anything meaningful. I would rather attempt to improve TestApp to allow upgrades and reuse existing tests (similar to the nondeterminism check Luke added) rather than try to get full coverage in container test.